File: inspect-db-formulas.js
// Direct DB inspection for formula duplication bug
const Database = require('better-sqlite3');
const db = new Database('./datos/panel.sqlite', { readonly: true });

// Check all modules for mixed account+section formulas
const rows = db.prepare(`
  SELECT DISTINCT empresa_id, modulo, anio, capitulo, clase, operacion_etiqueta, formula_json
  FROM layout_operaciones
  WHERE formula_json IS NOT NULL AND formula_json != '' AND formula_json != '[]'
  AND empresa_id = 'empresa1'
  ORDER BY anio DESC, modulo, capitulo, clase
`).all();

console.log(`Total rows with formula_json: ${rows.length}`);

const seen = new Set();
const issues = [];

rows.forEach(r => {
  const key = `${r.modulo}|${r.anio}|${r.capitulo}|${r.clase}`;
  if (seen.has(key)) return; // dedupe - same formula across tipo rows
  seen.add(key);

  let parsed = null;
  try { parsed = JSON.parse(r.formula_json); } catch (e) { return; }
  if (!parsed || !Array.isArray(parsed.tokens)) return;

  const refs = parsed.tokens.filter(t => t.kind === 'ref');
  const types = refs.map(t => t.refType);
  const uniqueTypes = [...new Set(types)];
  const hasMixed = types.includes('account') && (types.includes('section') || types.includes('subsection'));

  if (hasMixed) {
    issues.push({
      modulo: r.modulo,
      anio: r.anio,
      capitulo: r.capitulo,
      clase: r.clase.substring(0, 50),
      refTypes: uniqueTypes,
      tokens: refs.map(t => ({ refType: t.refType, label: t.label || t.refId }))
    });
  }
});

if (issues.length === 0) {
  console.log('\n✓ No formulas with mixed account+section refs found in DB.');
  console.log('=> The duplication is NOT in stored data, it happens during load/rendering.');
} else {
  console.log(`\n⚠ Found ${issues.length} operations with MIXED account+section refs:\n`);
  issues.forEach(i => console.log(JSON.stringify(i, null, 2)));
}

// Additionally show a sample of formula_json to understand their structure
console.log('\n=== SAMPLE FORMULAS (first 10 unique) ===');
const sampleSeen = new Set();
let count = 0;
rows.forEach(r => {
  const key = `${r.modulo}|${r.anio}|${r.capitulo}|${r.clase}`;
  if (sampleSeen.has(key) || count >= 10) return;
  sampleSeen.add(key);
  count++;
  let parsed = null;
  try { parsed = JSON.parse(r.formula_json); } catch (e) {}
  const refs = parsed && Array.isArray(parsed.tokens) ? parsed.tokens.filter(t => t.kind === 'ref') : [];
  console.log(`[${r.modulo}/${r.anio}/${r.capitulo}] ${r.clase.substring(0,40)}`);
  console.log(`  refTypes: [${[...new Set(refs.map(t => t.refType))].join(', ')}]`);
  console.log(`  labels: [${refs.map(t => t.label || t.refId).join(', ')}]`);
  console.log();
});

db.close();

File: inspect-db-formulas.py
import sqlite3, json

db = sqlite3.connect('./datos/panel.sqlite')
c = db.cursor()

c.execute('''
  SELECT DISTINCT modulo, anio, capitulo, clase, formula_json
  FROM layout_operaciones
  WHERE formula_json IS NOT NULL AND formula_json != '' AND formula_json != "[]"
  AND empresa_id = "empresa1"
  ORDER BY anio DESC, modulo, capitulo, clase
''')

rows = c.fetchall()
print(f'Total unique formula rows: {len(rows)}')

issues = []
samples = []

for modulo, anio, capitulo, clase, formula_json in rows:
    try:
        parsed = json.loads(formula_json)
    except:
        continue
    if not isinstance(parsed, dict) or not isinstance(parsed.get('tokens'), list):
        continue
    refs = [t for t in parsed['tokens'] if t.get('kind') == 'ref']
    ref_types = list(set(t.get('refType', '') for t in refs))
    has_mixed = 'account' in ref_types and ('section' in ref_types or 'subsection' in ref_types)
    labels = [t.get('label', t.get('refId', '')) for t in refs[:8]]
    if has_mixed:
        issues.append({
            'modulo': modulo, 'anio': anio, 'capitulo': capitulo,
            'clase': clase[:50], 'refTypes': ref_types, 'labels': labels
        })
    if len(samples) < 10:
        samples.append({
            'modulo': modulo, 'anio': anio, 'capitulo': capitulo,
            'clase': clase[:50], 'refTypes': ref_types
        })

if not issues:
    print('\nOK: No mixed account+section formulas found in DB.')
    print('=> Corruption happens AFTER loading (in frontend rendering/dedup).')
else:
    print(f'\nBUG: {len(issues)} operations with mixed account+section refs:')
    for i in issues[:10]:
        print(json.dumps(i, ensure_ascii=False))

print('\n=== SAMPLE FORMULAS ===')
for s in samples:
    modulo_s = s['modulo']
    anio_s = s['anio']
    capitulo_s = s['capitulo']
    clase_s = s['clase']
    reftypes_s = s['refTypes']
    print(f'  [{modulo_s}/{anio_s}/{capitulo_s}] {clase_s} => {reftypes_s}')

db.close()
